src/Language/Fixpoint/Solver/Interpreter.hs

```diff
@@ -0,0 +1,887 @@
+    go (PAll _ _)      = []
+    go (PExist _ _)    = []
+    go (PGrad{})       = []
+
+
+fastEval :: ConstMap -> Knowledge -> ICtx -> InstRes -> Expr -> EvalST Expr
+fastEval ienv γ ctx res e 
+    = do env  <- gets (seSort . evEnv)
+         return $ mytracepp ("evaluating" ++ show e) $ interpret ienv γ ctx env res $ simplify γ ctx e
+
+--------------------------------------------------------------------------------
+-- | 'substEq' unfolds or instantiates an equation at a particular list of
+--   argument values. We must also substitute the sort-variables that appear
+--   as coercions. See tests/proof/ple1.fq
+--------------------------------------------------------------------------------
+
+unfoldExpr :: ConstMap -> Knowledge -> ICtx -> SEnv Sort -> InstRes -> Expr -> {-EvalST-} Expr
+unfoldExpr ie γ ctx env res (EIte e0 e1 e2) = let g' = interpret' ie γ ctx env res e0 in
+                                                 if g' == PTrue
+                                                    then unfoldExpr ie γ ctx env res e1
+                                                    else if g' == PFalse
+                                                            then unfoldExpr ie γ ctx env res e2
+                                                            else EIte g' e1 e2
+unfoldExpr _  _ _   _   _   e               = e
+
+substEq :: SEnv Sort -> Equation -> [Expr] -> Expr
+substEq env eq es = subst su (substEqCoerce env eq es)
+  where su = mkSubst $ zip (eqArgNames eq) es
+
+substEqCoerce :: SEnv Sort -> Equation -> [Expr] -> Expr
+substEqCoerce env eq es = Vis.applyCoSub coSub $ eqBody eq
+  where 
+    ts    = snd    <$> eqArgs eq
+    sp    = panicSpan "mkCoSub"
+    eTs   = sortExpr sp env <$> es
+    coSub = mkCoSub env eTs ts
+
+mkCoSub :: SEnv Sort -> [Sort] -> [Sort] -> Vis.CoSub
+mkCoSub env eTs xTs = M.fromList [ (x, unite ys) | (x, ys) <- Misc.groupList xys ] 
+  where
+    unite ts    = Mb.fromMaybe (uError ts) (unifyTo1 senv ts)
+    senv        = mkSearchEnv env
+    uError ts   = panic ("mkCoSub: cannot build CoSub for " ++ showpp xys ++ " cannot unify " ++ showpp ts) 
+    xys         = Misc.sortNub $ concat $ zipWith matchSorts _xTs _eTs
+    (_xTs,_eTs) = (xTs, eTs)
+
+matchSorts :: Sort -> Sort -> [(Symbol, Sort)]
+matchSorts s1 s2 = go s1 s2
+  where
+    go (FObj x)      {-FObj-} y    = [(x, y)]
+    go (FAbs _ t1)   (FAbs _ t2)   = go t1 t2
+    go (FFunc s1 t1) (FFunc s2 t2) = go s1 s2 ++ go t1 t2
+    go (FApp s1 t1)  (FApp s2 t2)  = go s1 s2 ++ go t1 t2
+    go _             _             = []
+
+--------------------------------------------------------------------------------
+
+eqArgNames :: Equation -> [Symbol]
+eqArgNames = map fst . eqArgs
+
+interpret' :: ConstMap -> Knowledge -> ICtx -> SEnv Sort -> InstRes -> Expr -> Expr
+interpret' ie γ ctx env res e = mytracepp ("Interpreting " ++ show e) $ interpret ie γ ctx env res e
+
+interpret :: ConstMap -> Knowledge -> ICtx -> SEnv Sort -> InstRes -> Expr -> Expr
```

Comment:
Some documentation here explaining how the expression is transformed would help. In particular, I'm curious about the role of the `InstRes` parameter, but also how much it leaves unevaluated.